### ISSUE TITLE
White horizontal straps on print-outs

### ISSUE THREAD

**Issue by petrasteri** [sumatrapdfreader/sumatrapdf#919]
I've come across a printing issue in SumatraPDF 3.1.2 (32-bit and 64-bit).  Print-outs have horizontal straps at regular intervals.  I'll upload a sample PDF and a photo showing the issue.  The sample PDF is just a simple Excel file saved as PDF file.  I would get the same result using any PDF file.

This only happens to printer drivers of Xerox WorkCentre 7225 and 7855.  But I don't have this issue with driver of other Xerox printers.  The WorkCentre 7225 driver version is "5.295 PCL6".  I could provide the driver setup if this could help resolve the issue.

Sample PDF file:
[white_straps_on_print_outs_source.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/1412453/white_straps_on_print_outs_source.pdf)

Photo showing print-out:
![white_straps_on_print_outs_result](https://user-images.githubusercontent.com/12992229/31967090-1fe56432-b90d-11e7-993b-3b65d8bb5ff4.jpg)

**Comment by SumatraPeter** [comment]
You can try [toggling the "Print as Image"](https://github.com/sumatrapdfreader/sumatrapdf/issues/329) advanced printing option in [v3.0](https://www.sumatrapdfreader.org/download-prev.html) and see if it helps. Unfortunately this option was [removed in v3.1](https://www.sumatrapdfreader.org/news.html).

**Comment by petrasteri** [comment]
Unfortunately, I cannot use v 3.0 because of the bug https://github.com/sumatrapdfreader/sumatrapdf/issues/40 which appeared at this version but corrected at v 3.1

I have not reported on their forum because this issue seems to be more of SumatraPDF than of Xerox, because if I use **Acrobat Reader**, **Foxit Reader** or **PDF X-Change Viewer**, print-outs are good.

**Comment by SumatraPeter** [comment]
Have you tried the latest pre-release version instead of 3.1.2?

**Comment by petrasteri** [comment]
No, I have not tried pre-release because there is no obvious link in the website, so I suppose it's not for users to try them.
If you could give me the link, I could try.

**Comment by SumatraPeter** [comment]
https://www.sumatrapdfreader.org/docs/SumatraPDF-documentation-fed36a5624d443fe9f7be0e410ecd715.html

**Comment by petrasteri** [comment]
Pre-release download link is written inside "Documentation" page!  =_='

**Comment by petrasteri** [comment]
Just tested SumatraPDF-prerelease-10740-64-install.exe but unfortunately, the issue is still present.

**Comment by SumatraPeter** [comment]
Can't think of anything else you can try, so good luck and hope the Sumatra dev can solve your problem if the app is indeed at fault.

**Comment by petrasteri** [comment]
After some search on Internet, I found that this bug has been reported before.  Here are some links as examples.

https://github.com/sumatrapdfreader/sumatrapdf/issues/417
In this bug, the issue appears with HP and Ricoh printers too.
There's a suggested workaround.  I'll try that.

https://github.com/sumatrapdfreader/sumatrapdf/issues/685
Issues on Sharp printers are reported.

http://forums.fofou.org/sumatrapdf/topic?id=3185289

**Comment by SumatraPeter** [comment]
The suggested workaround seems to be related to the same _PrintAsImage_ option in v3.0 that I mentioned above, but since you can't use v3.0...

**Comment by petrasteri** [comment]
The workaround is
Printing Preferences -> Advanced tab -> Driver: disable "Print optimization"

Luckily, this works for me when I print on Xerox 7855.

**Comment by SumatraPeter** [comment]
Aah, so it _is_ something that particular Xerox model's driver is configured to do after all. Probably the other models aren't similarly configured by default (or someone changed the setting beforehand) which is why they have no problems with Sumatra.

Since Sumatra can't be expected to work around the quirks (and custom settings) of every printer driver out there, I fail to see the point of keeping this issue open.

**Comment by petrasteri** [comment]
What!?  Somebody found a workaround for a bug in SumatraPDF and you accuse that workaround as culprit?  What kind of logic is that?

**Comment by SumatraPeter** [comment]
Really? How do you know it's not other programs working around a bug in this so-called "Print optimization" that just some printer drivers are performing? All I'm saying is that it's far from conclusively proven that Sumatra's at fault here. Hopefully @kjk can investigate and identify the culprit once and for all, and naturally fix the bug if it's Sumatra. But if it's not then naturally it would be pointless keeping this issue open.

**Comment by Zeebrommer** [comment]
As a long time SumatraPDF user I have been encountering this issue on and off with Xerox printers (Xerox PullPrn PS V4.0.548.8.0) at work too. My workaround is to print the same document, with the same printer driver but using Adobe reader.

**Comment by weshatheleopard** [comment]
Having the same issue here.

**Comment by petrasteri** [comment]
> Really? How do you know it's not other programs working around a bug in this so-called "Print optimization" that only certain printer drivers are performing?

Sure!  Because I could return the same question to you: how can you know it's NOT sumatrapdf the source of the problem.

However, by Occam's razor, it's more **likely** that the problem lies in sumatrapdf than the other way round.

> All I'm saying is that it's far from conclusively proven that Sumatra's at fault here.

Once again, by Occam's razor.

> Hopefully @kjk can investigate and identify the culprit once and for all, and naturally fix the bug if it's Sumatra. But if it's not then of course it would be pointless keeping this issue open.

I hope the bug would be fixed soon, too, because I have stopped using SumatraPDF since my last comment, ie end of **2017**.

But by looking at the number of incidents open in GitHub on this similar issue, this bug really needs to be fixed ASAP.

Why not just revert the code for this part to version 3.0?

**Comment by GitHubRulesOK** [comment]
@petrasteri Not wishing to disappoint however since SumatraPDF is positioned as a Reader with support for basic system printing (not a virtual print driver) this issue is unlikely to be resolved unless someone can devote a considerable amount of time and even then the result is unlikely to please everyone. 
Your key point is you wish to revert SumatraPDF to the pre 3.0 print functions  which were themselves causing so many issues they were dropped. You point out there are many other good printing applications available and I am struggling as to why it is "essential" that SumatraPDF join them since that was not its aim to be a heavy weight application simply an enhanced lightweight screen viewer (currently providing a number of user interfaces to MuPDF)

Always trying to help
IF earlier works for print output but there is an issue with a certain type of file you give as example type #40 which was buggy from 2.5 to 3.0) then either

A) revert to version 2.4 for printing the "old way"
B) use latest pre-release etc to first print to Windows PDF then send that to version 3.0

**Comment by svaningelgem** [comment]
Hello everyone, as I am also impacted by this issue, I decided to dig into it.
I am quite vexed by the problem.
Short (I hope) summary of my investigation till now:
1) release 3.0 32bit: I can print the page normally.
2) git: can't print normally anymore.

But I made the 3.0 experience the exact same issues (so I'm guessing this proves that the issue is with this pdf reader) by modifying the following line:
`Print.cpp @ +- line 309:
if (false && !pd.advData.asImage) {
`
I added the false there to not go into this if-branch, but go to the else-branch, which is identical to how it is in git.

I outputted the generated HBITMAP to a file so I can have a look at it, and it looks just fine (also printing to PDFCreator 3.3.2 works fine without the horizontal stripes).
But it seems that one you do the "bmp->StretchDIBits" it distorts it somehow (but only for certain printer drivers?? -- as I said output to PDFCreator works fine).

That's what's vexing me.


Another thing:
I printed 2 pdf's. 1 pdf has 27 horizontal stripes on it (portrait pdf page), and the other only has 7 (landscape mode).

Next:
Starting from the 3.0 as I can replicate the behaviour, and I'm thinking it has to do something with the stretching, I went onto trying to stretch things differently.
1. Instead of
`RectI rc(offset.x, offset.y, bmp->Size().dx * shrink, bmp->Size().dy * shrink);`
I tried:
`RectI rc = RectI::FromXY( offset.x, offset.y, paperSize.dx, paperSize.dy );`
==> So this is the same as in the if-branch.
==> Result: nicely printed (meaning no stripes anymore), just the right and bottom margin are really bad. They're much too close to the edge of the paper.

Conclusion: it definitly has something to do with the stretching, and the target size!
What... That I don't understand (yet).


I tried in PDFEngine(2230) to use the if "PreferGdiPlusDevice" branch, but that results in the same thing: stripes.

I also noticed that in pdfium they use StretchDIBits instead of StretchBlt to draw to the HDC.


So I hope this would be of some help to someone, as I'm out of ideas. Maybe changing the code to use StretchDIBits might result in something good?
But why does the previous code of "engine.RenderPage" works fine, whereas "engine.RenderBitmap" + "bmp->StretchDIBits(dc)" produces the wrong output?


Both eventually end up in RunPage, so I took a snapshot there of the differences:
- cliprect @ RenderPage: x/y = -120 vs @ RenderBitmap: x/y = 0 (but x1 & y1 are the same)
- ctm @RenderPage is different too, but that's likely because ctm->e = -120, so ctm->f is also 120 less (probably because of the previous differences)
- all functions use the _gdiplus_ variant (RenderPage) instead of the _draw_ variant (RenderBitmap).


So right now I'm giving up on this issue. I'll need to look for another pdf viewer sadly now that opens fast (or go back to 3.0).

**Comment by GitHubRulesOK** [comment]
@kjk
can you please checkout/comment on above analysis from a user with hardware for testing https://github.com/sumatrapdfreader/sumatrapdf/issues/919#issuecomment-453781853 it seems to suggest a possible cause for some users problems

**Comment by rawtaz** [comment]
This sure seems like the closest to finding the cause anyone has gotten, and it seems like a pretty solid analysis. Any way you can look into this @kjk? Please see the comment linked to earlier, namely [this](https://github.com/sumatrapdfreader/sumatrapdf/issues/919#issuecomment-453781853).

FWIW I'm having the same issue as everyone else in this regard. Printing to a Xerox printer.

**Comment by SumatraPeter** [comment]
@rawtaz: Have you tested with the latest [daily build](https://www.sumatrapdfreader.org/dailybuilds.html) yet?

**Comment by rawtaz** [comment]
@SumatraPeter No, I haven't. I can coordinate with the affected user(s) so that they will do that if needed. But seeing as this issue has been running for years, is there anything in the daily build that makes you think that things will be different with it compared to version 3.2.0.0? If not I fail to see the point in trying the daily build, but I'm open to reasons for doing so of course.

**Comment by SumatraPeter** [comment]
@rawtaz: Was just a suggestion since kjk has been keeping the underlying MuPDF engine updated as of late, so it _may_ have improved matters. Your users can try a portable copy if they wish so that their existing installed version remains untouched. Or not. \*shrug*

**Comment by svaningelgem** [comment]
@SumatraPeter : I just tried it with my installed version (3.0). Works fine.
Afterwards I tried it with `Latest SumatraPDF daily build, built on 2020-08-12, commit 3e1de028f81f4c68e570f701022cd028c8041824`: horizontal stripes (1 about every cm) in a portrait page. So no change yet.

**Comment by rawtaz** [comment]
QUESTION: Can't we just bring back the "Print as image" option? That would work as a workaround for most if not all users encountering this problem, I think?

**Comment by SumatraPeter** [comment]
@rawtaz: [I wish](https://github.com/sumatrapdfreader/sumatrapdf/issues/329#issuecomment-152215551)... 🙁

**Comment by priyeshshah11** [comment]
@kjk I am still facing this issue, was there a conclusion on this or do we know where it's coming from? What did others do to fix it or get around it?

**Comment by svaningelgem** [comment]
I reverted to v3.0. Works fine.

**Comment by priyeshshah11** [comment]
> I reverted to v3.0. Works fine.

Thanks @svaningelgem, which version were you using previously where you were facing this error?

**Comment by svaningelgem** [comment]
See above in this thread 😁

**Comment by svaningelgem** [comment]
https://github.com/sumatrapdfreader/sumatrapdf/issues/919#issuecomment-673962651: 3.0 & latest git build at the time.

**Comment by priyeshshah11** [comment]
I still don't understand what you mean, because [you said](https://github.com/sumatrapdfreader/sumatrapdf/issues/919#issuecomment-1449357493) `"I reverted to v3.0. Works fine."` & [my question](https://github.com/sumatrapdfreader/sumatrapdf/issues/919#issuecomment-1449359157) was/is which version did you revert from? i.e. on which build it wasn't working for you if it is working on v3.0. 

I understood the git build part but was wondering which version was that on, I was then able to find it by finding that [commit](https://github.com/sumatrapdfreader/sumatrapdf/commit/3e1de028f81f4c68e570f701022cd028c8041824) & it was [v3.4.6](https://github.com/sumatrapdfreader/sumatrapdf/releases/tag/3.4.6rel).

Also, just for future reference it's always good to link the commits in the comments than posting the hash, easier for people to find. Thanks for all your help! 😄

**Comment by petrasteri** [comment]
> @kjk [snipped...] What did others do to fix it or get around it?

OP here.  I stopped using SumatraPDF LOL
I'm no longer interested in this bug.  But since I'm the OP, I suppose I have to leave this bug open and continue to receive updates from it.
Could I pass the "relay baton" to someone else?

**Comment by priyeshshah11** [comment]
You could just unsubscribe from this issue, right? if you don't want to receive these notifications anymore? <img width="317" alt="Screen Shot 2023-03-01 at 11 02 48 PM" src="https://user-images.githubusercontent.com/38547776/222134094-3dd74c32-d74e-4a8b-bf94-38d384a38742.png">

**Comment by weshatheleopard** [comment]
@petrasteri 

> I stopped using SumatraPDF LOL

I stopped using the printer that experienced these issues LOO

**Comment by Polynator** [comment]
Dear all,

I am facing this same issue ever since 3.1 was released back in 2015, and therefore am still at 3.0.

I am fully aware this is no paid show, but I assume there are others out there too, who would very much appreciate all the great progress of the recent versions without loosing the possibility to successfully print the occasional page.
It would make me and hopefully lots of others very, very happy if this problem could be finally solved.

Unfortunately, I am technically unable to contribute anything meaningful to this besides the following observations:
- The issues exists in every newer version including the current 3.5.15481
- If it helps I have attached another file that reliably fails.
- [Pflegeanleitung-Hydropflanzen_200526.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/12162686/Pflegeanleitung-Hydropflanzen_200526.pdf)

Best regards,
David

**Comment by weshatheleopard** [comment]
@Polynator As sad as it is, it looks like the problem is specific to a particular printer model(s). It may be an unfortunate combination of bugs both in SumatraPDF *and* the printer firmware, where either of these do not cause the issue by themselves, but when brought together, they do — that's called "synergy". The quickest solution is to use a different printer. Sorry that I couldn't help further.

**Comment by Polynator** [comment]
@weshatheleopard Thanks anyway.
Unfortunately, the replacement of fully-optioned machines in the range of 2k€ each is no very attractive option.

As far as I understand, the old printing option was unstable for some users. What I don't understand is, why it could not be left in the software, but deactivated by default.
On the other hand, @svaningelgem seems to have been on a good way to a solution. He even managed to print successfully only with badly adjusted right and bottom margin. (Which seems to me, as a layman, to be a solvable thing)

Hope dies last ;-)

Best regards,
David

**Comment by sebma** [comment]
I have the same problem on Xerox C8155 and I see no `Print optimization` to disable under Printing Preferences -> Advanced tab -> Driver :

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/4569306/3df89a67-fb5b-4711-9634-52d25dd42d4d)

**Comment by petrasteri** [comment]
> @petrasteri
> 
> > I stopped using SumatraPDF LOL
> 
> I stopped using the printer that experienced these issues LOL

Well, if you use some cheap printers, of course your "solution" is working.  But in a professional/business environment that I doubt is in your case, printers are very expensive -- can go up to $1000+.  In such case, it would be a lot MORE cost-effective to change the free software than changing the printer.

**Comment by GitHubRulesOK** [comment]
Long established PDF File Printer Software developers are Adobe and Artifex (who supply SumatraPDF core code (MuPDF) AND GhostScript.

One is designed for rendering images (has limited print abilities) the other has limited viewing but very well designed for printing !

**Comment by sebma** [comment]
@giveitacatch Nope, I wish I knew which option to use.

**Comment by kjk** [comment]
likely outside of sumatra control

**Comment by svaningelgem** [comment]
I actually beg to differ @kjk . The reason is that in a previous version it worked just fine (which is the version I'm still using).

And I understand where you're coming from as it is the way Sumatra is accessing the driver.

But I'm not knowledgeable enough to fix it...

**Comment by svaningelgem** [comment]
Thanks @kjk , I'll have a look when the new release is out to see if it has been resolved now. Love your product!
